clang/include/clang/AST/PreorderAST.h

```diff
@@ -0,0 +1,165 @@
+//===------- PreorderAST.h: An n-ary preorder abstract syntax tree -------===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+//  This file defines the interface for an n-ary preorder abstract syntax tree
+//  which is used to semantically compare two expressions.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_CLANG_PREORDER_AST_H
+#define LLVM_CLANG_PREORDER_AST_H
+
+#include "clang/AST/ASTContext.h"
+#include "clang/AST/CanonBounds.h"
+#include "clang/AST/Expr.h"
+
+namespace clang {
+  // Each node of the PreorderAST has a variable-count list which stores the
+  // Name of each variable along with a count of the number of times the
+  // variable appears in the node.
+  struct VarTy {
+    std::string Name;
```

Comment:
It would be better to use a VarDecl instead.   Because of the way bounds expressions work, you could have occurrences of two variables in the same scope with the same name.  Consider
```
array_ptr<int> p : count(len) = ...
int len = ...
... *p ...  // has a bounds expression that uses the first occurrence of len.
```